src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryRegEx.cs

```diff
@@ -0,0 +1,277 @@
+        {
+            Dictionary<string, NamedValue> fields = new ()
+                {
+                    { FULLMATCH, new NamedValue(FULLMATCH, StringValue.New(m.Value)) },
+                    { STARTMATCH, new NamedValue(STARTMATCH, NumberValue.New(m.Index + 1)) }
+                };
+
+            List<string> subMatches = new List<string>();
+            string[] groupNames = rex.GetGroupNames();
+
+            for (int i = 0; i < groupNames.Length; i++)
+            {
+                string groupName = groupNames[i];
+                string validName = DName.MakeValid(groupName, out _).Value;
+                Group g = m.Groups[i];
+
+                if (!int.TryParse(groupName, out _))
+                {
+                    if (!fields.ContainsKey(validName))
+                    {
+                        fields.Add(validName, new NamedValue(validName, StringValue.New(g.Value)));
+                    }
+                    else
+                    {
+                        fields[validName] = new NamedValue(validName, StringValue.New(g.Value));
+                    }
+                }
+
+                if (i > 0)
+                {
+                    subMatches.Add(g.Value);
+                }
+            }
+
+            if (!fields.ContainsKey(SUBMATCHES))
+            {
+                fields.Add(SUBMATCHES, new NamedValue(SUBMATCHES, TableValue.NewSingleColumnTable(subMatches.Select(s => StringValue.New(s)).ToArray())));
+            }
+
+            return RecordValue.NewRecordFromFields(fields.Values);
+        }
+
+        private static DType GetRecordTypeFromRegularExpression(string regularExpression)
+        {
+            Dictionary<string, TypedName> propertyNames = new ();
+            Regex rex = new Regex(regularExpression);
+
+            propertyNames.Add(FULLMATCH, new TypedName(DType.String, new DName(FULLMATCH)));
+            propertyNames.Add(STARTMATCH, new TypedName(DType.Number, new DName(STARTMATCH)));
+            propertyNames.Add(SUBMATCHES, new TypedName(DType.CreateTable(new TypedName(DType.String, new DName(TexlFunction.ColumnName_ValueStr))), new DName(SUBMATCHES)));
+
+            foreach (string groupName in rex.GetGroupNames())
+            {
+                if (!int.TryParse(groupName, out _))
+                {
+                    DName validName = DName.MakeValid(groupName, out _);
+
+                    if (!propertyNames.ContainsKey(validName.Value))
+                    {
+                        propertyNames.Add(validName.Value, new TypedName(DType.String, validName));
+                    }
+                }
+            }
+
+            return DType.CreateRecord(propertyNames.Values);
+        }
+
+        internal class CommonMatchImplementation
+        {
+            protected internal static Task<FormulaValue> CommonMatchImpl(FormulaValue[] args, string defaultMatchOptions, Func<string, string, RegexOptions, FormulaValue> implementation)
```

Comment:
> Func<string, string, RegexOptions, FormulaValue> implementation [](http://example.com/codeflow?start=121&length=64)

Per comment above, instead of this callback, we could just have a protected virtual. 
We already have a base class relationship.